xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+    return delegate.awaitTermination(remainingTime, TimeUnit.NANOSECONDS);
+  }
+
+  @Override
+  public void awaitTermination() throws InterruptedException {
+    internalTerminationLatch.await();
+    delegate.awaitTermination();
+  }
+
+  @Override
+  public int getPort() {
+    return delegate.getPort();
+  }
+
+  @Override
+  public List<? extends SocketAddress> getListenSockets() {
+    return delegate.getListenSockets();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getServices() {
+    return delegate.getServices();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getImmutableServices() {
+    return delegate.getImmutableServices();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getMutableServices() {
+    return delegate.getMutableServices();
+  }
+
+  // Must run in SynchronizationContext
+  private void startDelegateServer() {
+    if (restartTimer != null && restartTimer.isPending()) {
+      return;
+    }
+    if (isServing) {
+      return;
+    }
+    if (delegate.isShutdown()) {
+      delegate = delegateBuilder.build();
+    }
+    try {
+      delegate.start();
+      listener.onServing();
+      isServing = true;
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(null);
+      }
+    } catch (IOException e) {
+      logger.log(Level.FINE, "Fail to start delegate server: {0}", e);
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(e);
+      }
+      restartTimer = syncContext.schedule(
+        new RestartTask(), retryDelayNano, TimeUnit.NANOSECONDS, timeService);
+    }
+  }
+
+  private final class RestartTask implements Runnable {
+    @Override
+    public void run() {
+      startDelegateServer();
+    }
+  }
+
+  @VisibleForTesting
+  void setTimeService(@Nullable ScheduledExecutorService scheduledExecutorService) {
+    if (scheduledExecutorService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    timeService = scheduledExecutorService;
+  }
+
+  private final class DiscoveryState implements LdsResourceWatcher {
+    private final String resourceName;
+    // Most recently discovered filter chains.
+    private List<FilterChain> filterChains = new ArrayList<>();
+    // Most recently discovered default filter chain.
+    @Nullable
+    private FilterChain defaultFilterChain;
+    private boolean stopped;
+
+    private DiscoveryState(String resourceName) {
+      this.resourceName = checkNotNull(resourceName, "resourceName");
+      xdsClient.watchLdsResource(resourceName, this);
+    }
+
+    @Override
+    public void onChanged(final LdsUpdate update) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          if (update.listener() == null) {
+            // Ignore updates not for servers.
```

Comment:
`update.listener().name` also to match `this.resourceName` ?

Reply:
The `update.listener() == null` check is really along the same lines. Why would you get an update "not for servers"?  It should really be a `checkNotNull` precondition check.